File: 952-word-subsets/word-subsets.cpp
class Solution {
public:
    vector<string> wordSubsets(vector<string>& words1, vector<string>& words2) {
        vector<string> ans;
      //  vector<unordered_map<char,int>> arr1(words1.size());
          vector<unordered_map<char,int>> arr2(words2.size());
          unordered_map<char,int> required ;
        
            // 
           

          for(int i=0;i<words2.size();i++){
             string str=words2[i];
            for(char ch: str){
                arr2[i][ch]++;
            }
          }

        for(auto it:arr2){
           
            for(auto val:it){
                required[val.first]=max(required[val.first],val.second);
              //   cout<<val.first<<"-"<<val.second<<"  ";
            }
            cout<<endl;
        }
           
        // for(auto it:required){
        //     cout<<it.first<<"-"<<it.second;
        //     cout<<endl;
        // }

         
        for(int i=0;i<words1.size();i++){

            unordered_map<char,int> map;
            for(auto ch:words1[i]){
                map[ch]++;
            }
            // for(auto it:map){
            //     cout<<it.first<<" "<<it.second<<" -- " ;
            // }
            cout<<endl;
            bool flag=true;

            for(auto val:required){
                if( val.second > map[val.first]){
                    flag=false;
                    break;
                }
                
                 
            }
            if(flag){
                    cout<<i;
                    ans.push_back(words1[i]);
                    }
        }
        
          return ans;
    }
};

File: 952-word-subsets/word-subsets.java
class Solution {
    public List<String> wordSubsets(String[] words1, String[] words2) {
        
        List<String> ans=new ArrayList<>();
         List<HashMap<Character,Integer>> arr1=new ArrayList<>();
           List<HashMap< Character,Integer>> arr2=new ArrayList<>();
         HashMap<Character,Integer> required =new HashMap<>();
       

        for (String word : words1) {
            HashMap<Character, Integer> tempMap = new HashMap<>();
            for (char ch : word.toCharArray()) {
                tempMap.put(ch, tempMap.getOrDefault(ch, 0) + 1);
            }
            arr1.add(tempMap);
        }

        for (String word : words2) {
            HashMap<Character, Integer> tempMap = new HashMap<>();
            for (char ch : word.toCharArray()) {
                tempMap.put(ch, tempMap.getOrDefault(ch, 0) + 1);
            }
            arr2.add(tempMap);
        }

        for(int i=0;i<arr2.size();i++){
            HashMap<Character,Integer> map=new HashMap<>(arr2.get(i));

            map.forEach((key,value)->{
                 required.put(key,Math.max(required.getOrDefault(key,0),value));
            });
           
        }
          for (Map<Character, Integer> map : arr2) {
            for (Map.Entry<Character, Integer> entry : map.entrySet()) {
                required.put(entry.getKey(), Math.max(required.getOrDefault(entry.getKey(), 0), entry.getValue()));
            }
        }


    for(int i=0;i<words1.length;i++){
       // AtomicBoolean flag=new AtomicBoolean(true);

        HashMap<Character, Integer> map =new HashMap<>(arr1.get(i));

        // map.forEach((key,value)->{
        //     if(required.getOrDefault(key,0)>value){
        //         flag.set(false);
        //         return;
        //     }
        // });
        // if(flag.get()==true)ans.add(words1[i]);

                final boolean[] flag = {true};  // Use an array to modify the flag inside forEach

            required.forEach((key, value) -> {
                if (value > map.getOrDefault(key, 0)) {
                    flag[0] = false;  // Update the flag
                }
            });

            if (flag[0]) {
                ans.add(words1[i]);
            }

     }

        
        return ans;
    }
}